vision/odlc_contour_filtering.py

```diff
@@ -0,0 +1,498 @@
+"""
+This file contains functions to filter contours to identify the odlc shapes from an already
+processed image. The only function that should be needed is filter_contour(), the rest are helper
+functions.
+"""
+import numpy as np
+from nptyping import NDArray, Shape, UInt8, IntC, Float32
+import cv2
+import vision.common.constants as consts
+import vision.common.bounding_box as bbox
+
+
+def test_heirarchy(hierarchy: consts.Hierarchy, contour_index: int) -> bool:
+    """
+    Function to test whether the contour at the given index contains another contour in it
+
+    Parameters
+    ----------
+    hierarchy : consts.Hierarchy
+        The 2nd value returned from cv2.findContours(), describes how contours are inside of
+        other contours
+    contour_index : int
+        The index of the contour to be examined (in the contours tuple and hierarchy array
+        returned from cv2.findContours())
+
+    Returns
+    -------
+    contains_contour : bool
+        True if the given contour has a contour inside of it, False otherwise
+    """
+    if hierarchy[0, contour_index, 2] <= 0:
+        return False
+    return True
+
+
+def test_min_area_box(contour: consts.Contour, max_box_ratio_range: float = 0.5) -> bool:
+    """
+    Will create a box around the given contour that has the smallest possible area
+    (not necessarily upright) and check if the box's aspect ratio is within the given range
+
+    Parameters
+    ----------
+    contour : consts.Contour
+        The individual contour to be evaluated (as returned from cv2.findContours)
+    max_box_ratio_range : float = 0.5
+        The maximum acceptable range of aspect ratios (centered on a 1:1 ratio), so if 0.5 is the
+        given parameter then the aspect ratio between the length and width must be between 0.5 and
+        1.5 (1 - 0.5 and 1 + 0.5). The default would correspond to a 2:1 or 1:2 aspect ratio range
+        the length/width vs. width/length does not matter
+
+    Returns
+    -------
+    acceptable_ratio : bool
+        Returns true if the aspect ratio of the min area box is inbetween the min and max
+    """
+    # NDArray[Shape["4, 2"], Float32] is the return type of cv2.boxPoints()
+    min_area_box: NDArray[Shape["4, 2"], Float32] = cv2.boxPoints(cv2.minAreaRect(contour))
+    # either length/width or width/length, does not matter
+    aspect_ratio: float = (cv2.norm(min_area_box[0] - min_area_box[1])) / (
+        cv2.norm(min_area_box[1] - min_area_box[2])
+    )
+    # pylint made me format the comparison like this
+    return 1 - max_box_ratio_range < aspect_ratio < 1 + max_box_ratio_range
```

Comment:
When I said comment, I meant on the github conversation. My bad. Don't need to include this comment in the code.